src/wizeng.main.v3

```diff
@@ -39,82 +39,77 @@ def main(args: Array<string>) -> int {
 		.exit(3);
 
-	if (!Strings.endsWith(path, ".wasm")) return ErrorBuilder.new()
-		.puts(path)
-		.puts(": unknown file format, expected .wasm\n")
-		.exit(4);
-
-	// Parse the binary module
-	var monitors = MonitorOptions.getMonitors();
-	if (monitors.length > 0) Execute.tiering.onMonitorsStart();
-	var result = engine.loadWasmFile(path);
-	var module: Module;
-	match (result) {
-		Ok(m) => {
-			module = m;
-			BasicTracing.instrumentModule(m);
-		}
-		FileNotFound => return ErrorBuilder.new()
-			.puts("wizeng: could not load file ")
+	var all_paths = Vector<string>.new();
+	if (WizengOptions.LINK_MODULES.val != null) {
+		all_paths = WizengOptions.LINK_MODULES.val;
+	}
+	all_paths.put(path);
+	var modules = Array<Module>.new(all_paths.length);
+	var names = Array<string>.new(all_paths.length);
+	for (i < all_paths.length) {
+		var path = all_paths[i];
+		if (!Strings.endsWith(path, ".wasm")) return ErrorBuilder.new()
 			.puts(path)
-			.ln()
-			.exit(2);
-		ParseError(code, path, section, error_index, error_pos, error_msg) => return ErrorBuilder.new()
-			.render(code, path, section, error_index, error_pos, error_msg)
-			.exit(3);
+			.puts(": unknown file format, expected .wasm\n")
+			.exit(4);
 	}
-	// Call monitors for parsed module
-	for (i < monitors.length) monitors[i].onParse(module);
-	var err = ErrorGen.new(path);
-	if (monitors.length > 0) Execute.tiering.onMonitorsFinish(module, err);
-
-	// Gather the list of import processors
-	var imports = Array<Exportable>.new(module.imports.length);
-	var i = Instantiator.new(engine.extensions, module, imports, err);
-	var map = Strings.newMap<ImportProcessor>();
-
-	for (j < module.imports.length) {
-		var name = module.imports[j].imp.module_name;
-		if (map[name] == null) {
-			var p = ModuleRegistry.get(name);
-			if (p == null) return ErrorBuilder.new()
+	var monitors = MonitorOptions.getMonitors();
+	if (monitors.length > 0) Execute.tiering.onMonitorsStart();
+	for (i < all_paths.length) {
+		var path = all_paths[i];
+		// Parse the binary module
+		var result = engine.loadWasmFile(path);
+		var module: Module;
+		match (result) {
+			Ok(m) => {
+				module = m;
+				BasicTracing.instrumentModule(m);
+			}
+			FileNotFound => return ErrorBuilder.new()
+				.puts("wizeng: could not load file ")
 				.puts(path)
-				.put2(": import[%d] from unknown module \"%s\"\n", j, name)
-				.exit(7);
-			var trace = BasicTracing.moduleFilter != null && BasicTracing.moduleFilter.matchesStr(name, null);
-			p.init(programArgs, trace, err);
-			map[name] = p;
-			i.processors.put(p);
+				.ln()
+				.exit(2);
+			ParseError(code, path, section, error_index, error_pos, error_msg) => return ErrorBuilder.new()
+				.render(code, path, section, error_index, error_pos, error_msg)
+				.exit(3);
 		}
+		modules[i] = module;
+		names[i] = extractName(path);
+	}
+	var err = ErrorGen.new(all_paths[all_paths.length - 1]);
+	// Call monitors for all modules
+	for (i < modules.length) {
```

Comment:
I like using the `for (x in array)` syntax whereever possible, and it looks like it will work here.